pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,153 @@
+	//      downtimeTolerance = lastRetry-retryPeriod == 78s
+	// 3. worst non-graceful lease acquisition is leaseDuration+retryPeriod == 163s
+	// 4. worst graceful lease acquisition is retryPeriod == 26s
+	if ret.LeaseDuration.Duration == 0 {
+		ret.LeaseDuration.Duration = 137 * time.Second
+	}
+
+	if ret.RenewDeadline.Duration == 0 {
+		// this gives 107/26=4 retries and allows for 137-107=30 seconds of clock skew
+		// if the kube-apiserver is unavailable for 60s starting just before t=26 (the first renew),
+		// then we will retry on 26s intervals until t=104 (kube-apiserver came back up at 86), and there will
+		// be 33 seconds of extra time before the lease is lost.
+		ret.RenewDeadline.Duration = 107 * time.Second
+	}
+	if ret.RetryPeriod.Duration == 0 {
+		ret.RetryPeriod.Duration = 26 * time.Second
+	}
+
+	retryTimes := int(math.Floor(float64(ret.RenewDeadline.Duration / ret.RetryPeriod.Duration)))
+	klog.Infof("The leader election gives %v retries and allows for %v of clock skew. The kube-apiserver downtime tolerance is %vs. Worst non-graceful lease acquisition is %v. Worst graceful lease acquisition is %v.",
+		retryTimes,
+		ret.LeaseDuration.Duration-ret.RenewDeadline.Duration,
+		(retryTimes-1)*(int(ret.RetryPeriod.Duration.Seconds())),
+		ret.LeaseDuration.Duration+ret.RetryPeriod.Duration,
+		ret.RetryPeriod,
+	)
+
+	if len(ret.Namespace) == 0 {
+		if len(defaultNamespace) > 0 {
+			ret.Namespace = defaultNamespace
+		} else {
+			// Fall back to the namespace associated with the service account token, if available
+			if data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
+				if ns := strings.TrimSpace(string(data)); len(ns) > 0 {
+					ret.Namespace = ns
+				}
+			}
+		}
+	}
+	if len(ret.Name) == 0 {
+		ret.Name = defaultName
+	}
+	return ret
+}
+
+func GetClusterInfraStatus(restClient *rest.Config) (*configv1.InfrastructureStatus, error) {
+	err := configv1.AddToScheme(scheme.Scheme)
+	if err != nil {
+		return nil, err
+	}
+	client, err := client.New(restClient, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+
+	infra := &configv1.Infrastructure{}
+	key := types.NamespacedName{Name: infraResourceName}
+	if err = client.Get(context.TODO(), key, infra); err != nil {
+		return nil, err
+	}
+
+	if infra == nil {
+		return nil, fmt.Errorf("getting resource Infrastructure (name: %s) succeeded but object was nil", infraResourceName)
+	}
+	return &infra.Status, nil
+
+}
+
+// LeaderElectionSNOConfig uses the formula derived in LeaderElectionDefaulting with increased
+// retry period and lease duration for SNO clusters that have limited resources.
+// This method does not respect the passed in LeaderElection config and the returned object will have values
+// that are overridden with SNO environments in mind.
+// This method should only be called when running in an SNO Cluster.
+func LeaderElectionSNOConfig(config configv1.LeaderElection) configv1.LeaderElection {
```

Comment:
Why don't we just import this? less code to maintain

https://pkg.go.dev/github.com/openshift/library-go@v0.0.0-20240205190119-68f38a5e48af/pkg/config/leaderelection#LeaderElectionSNOConfig